Examples/Google/SpeechToText/Sources/SpeechService.swift

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPC
+
+typealias Request = Google_Cloud_Speech_V1_StreamingRecognizeRequest
+typealias Response = Google_Cloud_Speech_V1_StreamingRecognizeResponse
+typealias StreamingRecognizeCall = BidirectionalStreamingCall
+
+// Track whether we are currently streaming or not
+enum State {
+  case idle
+  case streaming(StreamingRecognizeCall<Request, Response>)
+}
+
+final class SpeechService {
+  
+  // Generated SpeechClient for making calls
+  private var client: Google_Cloud_Speech_V1_SpeechClient
+  
+  // Track if we are streaming or not
+  private var state: State = .idle
+  
+  init() {
+    precondition(!Constants.apiKey.isEmpty, "Please refer to the README on how to configure your API Key properly.")
+    
+    // Make EventLoopGroup for the specific platform (NIOTSEventLoopGroup for iOS)
+    let group = PlatformSupport.makeEventLoopGroup(loopCount: 1)
+    
+    // Create connection channel with our group, host, and port
+    let channel = ClientConnection
+      .secure(group: group)
+      .connect(host: "speech.googleapis.com", port: 443)
+
+    // Specify call options to be used for gRPC calls
+    let callOptions = CallOptions(customMetadata: [
+      "x-goog-api-key": Constants.apiKey
+    ])
+
+    // Now we have a client!
+    self.client = Google_Cloud_Speech_V1_SpeechClient(channel: channel, defaultCallOptions: callOptions)
+  }
+  
+  func stream(_ data: Data,
+              completion: ((Google_Cloud_Speech_V1_StreamingRecognizeResponse) -> Void)? = nil) {
+
+    switch self.state {
+    case .idle:
+      // Initialize the bidirectional stream
+      let call = self.client.streamingRecognize { response in
+        // Message received from Server, execute provided closure from caller
+        completion?(response)
+      }
+
+      // Specify audio details
+      let config = Google_Cloud_Speech_V1_RecognitionConfig.with {
+        $0.encoding = .linear16
+        $0.sampleRateHertz = 16000
+        $0.languageCode = "en-US"
+        $0.enableAutomaticPunctuation = true
+        $0.metadata = Google_Cloud_Speech_V1_RecognitionMetadata.with {
+          $0.interactionType = .dictation
+          $0.microphoneDistance = .nearfield
+          $0.recordingDeviceType = .smartphone
+        }
+      }
+
+      // Create streaming request
+      let request = Google_Cloud_Speech_V1_StreamingRecognizeRequest.with {
+        $0.streamingConfig = Google_Cloud_Speech_V1_StreamingRecognitionConfig.with {
+          $0.config = config
+        }
+      }
+
+      // Send first message consisting of the streaming request details
+      call.sendMessage(request, promise: nil)
+
+      self.state = .streaming(call)
```

Comment:
Nit: can you move this to when you create `call`? It feels a bit out of place here since we: create a stream, send the initial configuration, update state, send data.